File: LeetCode/875_Koko_Eating_Bananas.js
var minEatingSpeed = function(piles, h) {
    var l = 1, r = Math.max(...piles)
    var result = r
    
    while (l <= r) {
      var k = Math.floor(l + r) / 2
      var hours = 0
      for (let p of piles) {
        hours += (p / k)
        // console.log(hours)
      }
      if (hours <= h) {
        result = Math.min(result, k)
        r = k - 1
      } else {
        l = k + 1
      }
    }
  return k
};



var piles = [3,6,7,11], h = 8
// var piles = [30,11,23,4,20], h = 5
// var piles = [30,11,23,4,20], h = 6

console.log(minEatingSpeed(piles, h))

File: LeetCode/875_Koko_Eating_Bananas.py
# Binary Search
import math
def minEatingSpeed(piles, h):
    l, r = 1, max(piles)
    result = r

    while l <= r:
        k = (l + r) // 2
        hours = 0
        for p in piles:
            hours += math.ceil(p / k)
        if hours <= h:
            result = min(result, k)
            r = k - 1
        else:
            l = k + 1
    return result



piles = [3,6,7,11]  
h = 8
piles = [30,11,23,4,20]  
h = 5
piles = [30,11,23,4,20]  
h = 6

print(minEatingSpeed(piles, h))